### PR TITLE
AC_PrecLand: rework estimation, add EST_TYPE param

### PR DESCRIPTION
**Pull request by jschall**
This allows the user to switch back to the old method of doing precision landing - using the raw measurements only. It also provides a framework for adding new estimators.

### CODE REVIEW

**Comment by rmackay9** [comment]
Thanks @jschall.  Has it been flight tested in the two configurations?

**Comment by jschall** [comment]
Not since the rebase...

File: libraries/AC_PrecLand/AC_PrecLand.cpp [hunk 1]
```diff
@@ -245,3 +189,158 @@ void AC_PrecLand::handle_msg(mavlink_message_t* msg)
...
+        _target_vel_rel_out_NE.x -= inertial_data.correctedVehicleDeltaVelocityNED.x;
+        _target_vel_rel_out_NE.y -= inertial_data.correctedVehicleDeltaVelocityNED.y;
+        _target_pos_rel_out_NE.x += _target_vel_rel_out_NE.x * inertial_data.dt;
+        _target_pos_rel_out_NE.y += _target_vel_rel_out_NE.y * inertial_data.dt;
+    }
+
+    // Apply offset
+    Vector3f land_ofs_ned_m = _ahrs.get_rotation_body_to_ned() * Vector3f(_land_ofs_cm_x,_land_ofs_cm_y,0) * 0.01f;
+    _target_pos_rel_est_NE.x += land_ofs_ned_m.x;
+    _target_pos_rel_est_NE.y += land_ofs_ned_m.y;
```
Comment (OXINARF) [inline libraries/AC_PrecLand/AC_PrecLand.cpp]: Shouldn't the offset be applied to the *_target_pos_rel\_**out**_NE*?

File: libraries/AC_PrecLand/AC_PrecLand.cpp [hunk 2]
```diff
@@ -50,7 +50,13 @@ const AP_Param::GroupInfo AC_PrecLand::var_info[] = {
     // @Units: Centimeters
     AP_GROUPINFO("LAND_OFS_Y",    4, AC_PrecLand, _land_ofs_cm_y, 0),
 
-    // 5 RESERVED for EKF_TYPE
+    // @Param: EST_TYPE
+    // @DisplayName: Precision Land Estimator Type
+    // @Description: Specifies the estimation method used.
+    // @Values: 0:RAW_MEASUREMENTS, 1:TWO_STATE_KF_PER_AXIS
```
Comment (GunMan-fr) [inline libraries/AC_PrecLand/AC_PrecLand.cpp]: Is it possible to get more info on which method would be better with what hardware ?

**Comment by fnoop** [comment]
Hi, I've reapplied these changes to current master:
https://github.com/fnoop/ardupilot/commit/a4b064ef764834daa7c528190c8072b78925e801
https://github.com/fnoop/ardupilot/commit/fab57a780c8d4d1908495df053e174bca06a5d0c

**Comment by fnoop** [comment]
So I've tested vision landing today using the raw estimator, and it works pretty well.  There's a really good correlation in sitl and real life between PL.pX and Pitch.  Using the ekf estimator in SITL there still seems to be quite an offset between PL.pX and Pitch.  I'll try and do a real life test using the ekf estimator later on.
I'm using a raspberry pi 2 (ie. very slow), sending landing_target messages between 5-10hz and with a fairly hefty latency (guessing >150ms).
But using the raw estimator, it's infinitely better than the current ekf-only code in master which flies off in random directions like a possessed wasp.

**Comment by rmackay9** [comment]
ok, this has finally reached the top of my list so looking at this today.

**Comment by rmackay9** [comment]
I've worked on this a bit today and here's some feedback:

- new branch rebased on master is here: https://github.com/rmackay9/rmackay9-ardupilot/tree/fnoop-precland
- the sensor based estimator has a bad jitter.  Logs are here: https://www.dropbox.com/s/do0ioht83ihetcv/new-fixed-no-param.bin?dl=0.  First part of the log is with the kalman-filter (worked a-ok!), 2nd part is with the sensor estimator which was jittery.
- I saw very very strange behaviour when I tried to add the ACC_P_NSE parameter (the last commit).  The "health" that was logged to the DF was "155" instead of just "1" and precision landing didn't work at all.  I'm investigating whether we've got parameter or memory corruption.

**Comment by rmackay9** [comment]
Closing this PR because I've opened a new one rebased on master.  https://github.com/ArduPilot/ardupilot/pull/5885

**Pull request closed without merging**